Source/Global/perform_env.cpp

```diff
@@ -315,58 +499,111 @@ HRESULT CALLBACK HC_PERFORM_ENV::CleanupAsyncProvider(XAsyncOp op, const XAsyncP
         }
+        return S_OK;
     }
-    case XAsyncOp::DoWork:
+    default:
     {
-        HC_UNIQUE_PTR<HC_PERFORM_ENV> reclaim{ env };
+        return S_OK;
+    }
+    }
+}
+
+void CALLBACK HC_PERFORM_ENV::ProviderCleanup(void* context, bool /*canceled*/)
+{
+    HC_UNIQUE_PTR<HC_PERFORM_ENV> env{ static_cast<HC_PERFORM_ENV*>(context) };
+    XAsyncBlock* cleanupAsyncBlock{ env->m_cleanupAsyncBlock };
+
 #if HC_PLATFORM == HC_PLATFORM_GDK
-        auto curlCleanupAsyncBlock = http_allocate_unique<XAsyncBlock>();
-        curlCleanupAsyncBlock->queue = data->async->queue;
-        curlCleanupAsyncBlock->context = data->async;
-        curlCleanupAsyncBlock->callback = [](XAsyncBlock* async)
-        {
-            HC_UNIQUE_PTR<XAsyncBlock> curlCleanupAsyncBlock{ async };
-            XAsyncBlock* envCleanupAsyncBlock = static_cast<XAsyncBlock*>(curlCleanupAsyncBlock->context);
-            
-            HRESULT cleanupResult = XAsyncGetStatus(curlCleanupAsyncBlock.get(), false);
-            curlCleanupAsyncBlock.reset();
-
-            // HC_PERFORM_ENV fully cleaned up at this point
-            XAsyncComplete(envCleanupAsyncBlock, cleanupResult, 0);
-        };
+    HC_UNIQUE_PTR<XAsyncBlock> curlCleanupAsyncBlock{ new (http_stl_allocator<XAsyncBlock>{}.allocate(1)) XAsyncBlock
+    {
+        cleanupAsyncBlock->queue,
+        cleanupAsyncBlock,
+        ProviderCleanupComplete
+    }};
 
-        auto curlProvider = std::move(env->curlProvider);       
-        reclaim.reset();
+    auto curlProvider = std::move(env->curlProvider);
+    env.reset();
 
-        RETURN_IF_FAILED(CurlProvider::CleanupAsync(std::move(curlProvider), curlCleanupAsyncBlock.get()));
+    HRESULT hr = CurlProvider::CleanupAsync(std::move(curlProvider), curlCleanupAsyncBlock.get());
+    if (FAILED(hr))
+    {
+        XAsyncComplete(cleanupAsyncBlock, hr, 0);
+    }
+    else
+    {
         curlCleanupAsyncBlock.release();
-        return E_PENDING;
+    }
 #else
-        reclaim.reset();
-        XAsyncComplete(data->async, S_OK, 0);
-        return S_OK;
+    // No additional provider cleanup needed
+    env.reset();
+    XAsyncComplete(cleanupAsyncBlock, S_OK, 0);
 #endif
+}
+
+void CALLBACK HC_PERFORM_ENV::ProviderCleanupComplete(XAsyncBlock* async)
+{
+#if HC_PLATFORM == HC_PLATFORM_GDK
+    HC_UNIQUE_PTR<XAsyncBlock> curlCleanupAsyncBlock{ async };
+    XAsyncBlock* envCleanupAsyncBlock = static_cast<XAsyncBlock*>(curlCleanupAsyncBlock->context);
+
+    HRESULT cleanupResult = XAsyncGetStatus(curlCleanupAsyncBlock.get(), false);
+    curlCleanupAsyncBlock.reset();
+
+    // HC_PERFORM_ENV fully cleaned up at this point
+    XAsyncComplete(envCleanupAsyncBlock, cleanupResult, 0);
+#else
+    UNREFERENCED_PARAMETER(async);
+    assert(false);
+#endif
+}
+
+bool HC_PERFORM_ENV::CanScheduleProviderCleanup()
+{
+    // mutex should always be held when calling this method
+    assert(!m_mutex.try_lock());
```

Comment:
don't do this, it will crash on some platforms
or at least wrap it in
```c++
#if !HC_PLATFORM_IS_EXTERNAL
```

Reply:
oh good to know. Will update

Reply:
is there a safe xplat way to assert that the mutex is held?

Reply:
no :(

Reply:
> If try_lock is called by a thread that already owns the mutex, the behavior is undefined.

Reply:
I'll just remove this altogether